### PR TITLE
Tuning rocsolver documentation

### PR DESCRIPTION
**Pull request by jzuniga-amd**
I am opening this as a draft for others to start reviewing it, but I still need to reword some parts of the functions where I have actually tuned the described constants during the optimization works.

### CODE REVIEW

**Review by tfalders — COMMENTED** [review]
Looking good so far. I have a few comments for you to look at, and have also opened a PR on your local fork with my suggested changes to the docs.

File: docs/Doxyfile [hunk 1]
```diff
@@ -772,7 +772,7 @@ WARN_LOGFILE           =
 # Possible values are: NO, YES and FAIL_ON_WARNINGS.
 # The default value is: NO.
 
-WARN_AS_ERROR          = YES
+WARN_AS_ERROR          = NO
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply docs/Doxyfile:775]: Yeah, I turned it off for debugging (to be able to build the docs while getting warnings). I'll be sure it is back on when the PR is ready.

File: library/src/include/rocblas.hpp [hunk 2]
```diff
@@ -14,6 +14,20 @@
 #include "lib_host_helpers.hpp"
 #include "rocsolver_logger.hpp"
 
+// THESE FOLLOWING VALUES ARE TO MATCH ROCBLAS C++ INTERFACE
+// THEY ARE DEFINED/TUNNED IN ROCBLAS
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/include/rocblas.hpp:18]: ok

File: library/src/auxiliary/rocauxiliary_sterf.hpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /************************************************************************
- * Derived from the BSD3-licensed
+ * Derived from the BS2D3-licensed
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/auxiliary/rocauxiliary_sterf.hpp:2]: Again, this was done by the clang-format script... not sure what is going on. @cgmb ?

File: library/src/lapack/roclapack_sytf2.hpp [hunk 4]
```diff
@@ -14,6 +14,10 @@
 #include "rocblas.hpp"
 #include "rocsolver.h"
 
+#define SYTF2_MAX_THDS                                    \
+    256 // thread-block size for calling the sytf2 kernel \
+        // MAX_THDS sizes must be one of 128, 256, 512, or 1024
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/lapack/roclapack_sytf2.hpp:19]: idem

File: library/src/lapack/roclapack_sytrf.hpp [hunk 5]
```diff
@@ -15,6 +15,10 @@
 #include "roclapack_sytf2.hpp"
 #include "rocsolver.h"
 
+#define SYTRF_MAX_THDS                                    \
+    256 // thread-block size for calling the sytrf kernel \
+        // MAX_THDS sizes must be one of 128, 256, 512, or 1024
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/lapack/roclapack_sytrf.hpp:20]: idem

File: library/src/include/ideal_sizes.hpp [hunk 6]
```diff
@@ -1,25 +1,211 @@
...
+
+/***************** sygs2/sygst and hegs2/hegst ********************************
+*******************************************************************************/
+/*! \brief Determines the size of the leading block that is reduced to standard form at each step
+    when using the blocked algorithm (SYGST/HEGST).
+
+    \details xxGST_BLOCKSIZE also acts as a switch size; if the original size of the problem is not larger than xxGST_BLOCKSIZE (n <= xxGST_BLOCKSIZE),
+    SYGST/HEGST will directly call the unblocked routines (SYGS2/HEGS2). However, when n is not
+    multiple of xxGST_BLOCKSIZE, the last block reduced in the blocked process is allowed to be smaller than xxGST_BLOCKSIZE. */
+#define xxGST_BLOCKSIZE 64
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/include/ideal_sizes.hpp]: Not a bad idea. I will add them to the functions and the documentation.

[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/include/ideal_sizes.hpp]: Actually, for sygst/hegst a switch size would have no effect as the algorithm combines right-looking and left-looking (for problems types 2 and 3), ie. the matrix update should always happen before the last block reduction anyways. 

For potrf it is OK. I added the switch size as a separate constant.

**Comment by jzuniga-amd** [comment]
I am turning this into a normal PR now. I will leave the documentation of the optimized functions for another PR if we have the time. A number of constant names and scopes will change once I tune the npvt, batched and strided_batched routines, so no urgent need to document them right now anyways.

**Review by tfalders — APPROVED** [review]
Looks good to me!

**Comment by tfalders** [comment]
Some tests seem to be failing on CI, though. I don't know if it's related to this PR.

File: library/src/lapack/roclapack_potrf.hpp [hunk 7]
```diff
@@ -210,9 +212,16 @@ rocblas_status rocsolver_potrf_template(rocblas_handle handle,
...
                                          strideA, batch_count);
             }
+            j += nb;
         }
     }
 
+    // factor last block
+    if(j < n)
+        rocsolver_potf2_template<T>(handle, uplo, n-j, A, shiftA + idx2D(j, j, lda), lda,
+                                    strideA, iinfo, batch_count, scalars, (T*)work1, pivots);
```
Comment (tfalders) [inline library/src/lapack/roclapack_potrf.hpp:222]: Looks like this is the cause of the CI failures. POTF2 is putting its `info` in `iinfo`, but then `info` is not being updated using `chk_positive`.

File: library/src/auxiliary/rocauxiliary_sterf.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgmb) [reply library/src/auxiliary/rocauxiliary_sterf.hpp:2]: It doesn't do that for me. Are you on sure it wasn't changed when you were modifying the `BLOCKSIZE` macro?

File: library/src/auxiliary/rocauxiliary_lasyf.hpp [hunk 8]
```diff
@@ -13,6 +13,10 @@
 #include "rocblas.hpp"
 #include "rocsolver.h"
 
+#define LASYF_MAX_THDS                                    \
+    256 // thread-block size for calling the lasyf kernel \
+        // MAX_THDS sizes must be one of 128, 256, 512, or 1024
```
[the remark this answers is not in the document]
Comment (cgmb) [reply library/src/auxiliary/rocauxiliary_lasyf.hpp]: There might be a setting that could help, but your solution seems fine to me.

**Review by cgmb — COMMENTED** [review]
I gave it another editing pass in https://github.com/jzuniga-amd/rocSOLVER/pull/9.

One highlight is that I found why the table of contents looks weird. It seems we can fix it by updating the sphinx_rtd_theme. https://github.com/readthedocs/sphinx_rtd_theme/issues/931

File: library/src/auxiliary/rocauxiliary_laswp.hpp [hunk 9]
```diff
@@ -12,6 +12,8 @@
 #include "rocblas.hpp"
 #include "rocsolver.h"
 
+#define THDS 256 // size of thread-blocks for calling the laswp kernel
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/auxiliary/rocauxiliary_laswp.hpp:15]: That's exactly why, we are using the BLOCKSIZE name to refer to the size of submatrices analyzed at each step in blocked algorithms. `LASWP_BLOCKSIZE` was not the case, it is simply the number of threads in thread-blocks for launching the kernel.

[the remark this answers is not in the document]
Comment (cgmb) [reply library/src/auxiliary/rocauxiliary_laswp.hpp:15]: Ah. I see. It still might be nice to have a laswp-specific name, but that makes sense.

**Review by cgmb — APPROVED** [review]

**Review by tfalders — APPROVED** [review]

File: library/src/auxiliary/rocauxiliary_laswp.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/auxiliary/rocauxiliary_laswp.hpp:15]: Yeah, why not; I will call it `LASWP_THDS` as I did in `sytrf`, for example

**Pull request merged**